Fix: failed to parse `/*eslint` comments by colon (fixes #6224)

mysticatea opened the pull request:
Fixes #6224.

It came to use Optionator to parse the comments by the same way as parsing CLI option. But Optionator cannot parse commaless notation, so there is a fallback for that.

Also, I just reordered requirements by file paths.

mysticatea commented on lib/eslint.js:
```diff
@@ -780,7 +813,7 @@ module.exports = (function() {
                 var ruleCreator,
                     severity,
-                    options,
+                    ruleOptions,
```
Just avoided shadowing.

mysticatea commented on lib/eslint.js:
```diff
@@ -9,25 +9,25 @@
+    NodeEventGenerator = require("./util/node-event-generator"),
+    SourceCode = require("./util/source-code"),
+    Traverser = require("./util/traverser"),
+    options = require("./options"),
```
I added `options = require("./options"),` for this fix.

platinumazure commented:
Should we use `./options` rather than just pulling in levn or optionator directly? Using our wrapper seems like an unnecessary abstraction that just makes it harder to understand what's going on.

nzakas commented on lib/eslint.js:
```diff
@@ -70,6 +70,25 @@ function parseBooleanConfig(string, comment) {
     return items;
 }
 
+var VALID_LEVELS = [0, 1, 2, "off", "warn", "error"];
```
We use the term "severity" in most places, instead of "level".

nzakas commented on lib/eslint.js:
```diff
@@ -80,11 +99,26 @@ function parseBooleanConfig(string, comment) {
+
+        // Some tests say that it should ignore invalid comments such as `/*eslint no-alert:abc*/`.
+        if (levelsAreValid(items)) {
+            return items;
```
Does `else` fall through and reparse?

mysticatea replied:
Yes, I added comments to explain this.

nzakas commented on lib/eslint.js:
```diff
@@ -70,6 +70,25 @@ function parseBooleanConfig(string, comment) {
+ * @param {object} config - A config object to check.
+ * @returns {boolean} `true` if all rules' level is valid.
+ */
+function levelsAreValid(config) {
```
Maybe `isInlineConfigValid`?

Also, maybe create `ConfigOps.isValidSeverity()` to keep that logic in one place?

mysticatea commented:
- I changed `options.parse` to `levn.parse`. Levn is the parser Optionator is using, so this doesn't add any dependency.
- I moved a function to check severity to `config-ops.js`.

platinumazure commented on lib/config/config-ops.js:
```diff
@@ -248,6 +249,30 @@ module.exports = {
+     * @param {object} config - The configuration for rules.
+     * @returns {boolean} `true` if the configuration has valid severity.
+     */
+    isEveryValidSeverify: function(config) {
```
I think you might want `isEverySeverityValid` here.

mysticatea replied:
Thank you, I renamed it.
In fact, I was worried about this name :smile:

platinumazure commented on lib/eslint.js:
```diff
@@ -80,6 +80,25 @@ function parseBooleanConfig(string, comment) {
+        // Also, commaless notations have invalid severity:
+        //     "no-alert: 2 no-console: 2" --> {"no-alert": "2 no-console:2"}
+        // Should ignore that case as well.
+        if (ConfigOps.isEveryValidSeverify(items)) {
```
I realize this if statement shouldn't throw, but should this be outside the try block anyway just to be safe? (If the condition is prefixed with `items &&`, I think it can be done safely.)

mysticatea replied:
The target of this check is the result of `levn.parse`, so I think this should inside of this `try` block since this location will be executed only if `levn.parse` succeeded.

platinumazure commented:
Thanks! Really appreciate the switch to levn, it feels more semantically accurate. I threw in a couple of comments for some small questions I had, but they're really small things and overall this looks great.

nzakas commented:
Lgtm
